x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/JobStorageDeletionTask.java

```diff
@@ -106,6 +113,72 @@ public void delete(String jobId, Client client, ClusterState state,
+                        ML_ORIGIN,
+                        request,
+                        ActionListener.<AcknowledgedResponse>wrap(
+                            deleteIndexResponse -> deleteAliases(jobId, client, deleteAliasHandler), //skip DBQ
+                            failureHandler),
+                        client.admin().indices()::delete);
+                }
+            },
+            failureHandler
+        );
+
+        // Step 4(b). If we are shared, skip index deletion. If not, determine if we are the only job on the custom index
+        ActionListener<GetAliasesResponse> getAliasesHandler = ActionListener.wrap(
+            aliasesResponse -> {
+                boolean onlySingleJob = true;
+                for (AliasMetaData aliasMetaData : aliasesResponse.getAliases().getOrDefault(indexName, new ArrayList<>())) {
+                    onlySingleJob = onlySingleJob && aliasMetaData.alias().endsWith(jobId); // circuit break for performance
+                }
+                if (onlySingleJob) { // Could we just skip this search and simply rely on the index suffixes?
+                    SearchSourceBuilder source = new SearchSourceBuilder();
+                    BoolQueryBuilder builder = QueryBuilders.boolQuery().filter(
+                        QueryBuilders.boolQuery().mustNot(QueryBuilders.termQuery(Job.ID.getPreferredName(), jobId)));
+
+                    source.query(builder).size(1);
+
+                    SearchRequest searchRequest = new SearchRequest(indexName);
+                    searchRequest.source(source);
+                    executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, customIndexSearchHandler);
+                } else {
+                    deleteByQueryExecutor.onResponse(true);
+                }
+            },
+            failureHandler
+        );
+
+        // Step 4. Determine if we are on a shared index by looking at `.ml-anomalies-shared` or the custom index's aliases
+        ActionListener<Boolean> deleteCategorizerStateHandler = ActionListener.wrap(
+            response -> {
+                if (indexName.equals(AnomalyDetectorsIndexFields.RESULTS_INDEX_PREFIX +
+                    AnomalyDetectorsIndexFields.RESULTS_INDEX_DEFAULT)) {
+                    deleteByQueryExecutor.onResponse(true); //don't bother searching the index any further, we are on the default shared
+                } else {
+                    GetAliasesRequest aliasesRequest = new GetAliasesRequest();
+                    aliasesRequest.indices(indexName);
+                    executeAsyncWithOrigin(
+                        client.threadPool().getThreadContext(),
+                        ML_ORIGIN,
+                        aliasesRequest,
+                        getAliasesHandler,
+                        client.admin().indices()::getAliases);
+                }
+            },
+            failureHandler
+        );
```

Comment:
The callback style here is less than desired, I am going to see if I can clean it up further (ideas are welcome).

I could split up multiple methods, but that may cause more confusion and going straight through to DBQ is a common action.

Reply:
First, I think that the easiest way to follow those callbacks is to make the path straight. Each callback step is called every time in the same order. To achieve control flow, we handle the response at the first step of the callbacks. For example, if we don't need to do get aliases, we call `getAliasesHandler.onResponse(null)` and in the handler if the response is null we call the corresponding next step. This is open for discussion of course, but it feels keeping the path straight is less error prone than mixing handlers around.